chia/util/streamable.py

```diff
@@ -127,8 +132,55 @@ def recurse_jsonify(d):
 
 PARSE_FUNCTIONS_FOR_STREAMABLE_CLASS = {}
 
+_field_metadata_key = "_chia_streamable"
+
+
+@dataclasses.dataclass(frozen=True)
+class _FieldMetadata:
+    ignore: bool = False
+
+
+@functools.wraps(wrapped=dataclasses.field)
+def unstreamed_field(*args, **kwargs):
+    metadata = kwargs.setdefault("metadata", {})
+    metadata[_field_metadata_key] = _FieldMetadata(ignore=True)
+
+    return dataclasses.field(*args, **kwargs)
+
+
+_T_Field = TypeVar("_T_Field", bound="_Field")
+
+
+@dataclasses.dataclass(frozen=True)
+class _Field:
+    name: str
+    annotation: Any
+
+    @classmethod
+    def from_dataclass_field(cls: Type[_T_Field], field: dataclasses.Field) -> _T_Field:
+        return cls(name=field.name, annotation=field.type)
+
+
+_T_ClassMetadata = TypeVar("_T_ClassMetadata", bound="_ClassMetadata")
+
+
+@dataclasses.dataclass(frozen=True)
+class _ClassMetadata:
+    fields: Tuple[_Field, ...] = ()
+
+    @classmethod
+    def from_dataclass(cls: Type[_T_ClassMetadata], dataclass) -> _T_ClassMetadata:
+        fields = []
+        for dataclass_field in dataclasses.fields(dataclass):
+            metadata = dataclass_field.metadata.get(_field_metadata_key, _FieldMetadata())
+            if metadata.ignore:
+                continue
+            fields.append(_Field.from_dataclass_field(field=dataclass_field))
+
+        return cls(fields=tuple(fields))
+
 
-def streamable(cls: Any):
+def streamable(cls: Type["Streamable"]) -> Type["Streamable"]:
```

Comment:
Super-pedantic, but doesn't this technically take any class type and turn it into a `Streamable`? As in
```suggestion
def streamable(cls: Type["Any"]) -> Type["Streamable"]:
```

Reply:
Uh...  so it does...  This is all a bit odd what with the classes first inheriting from `Streamable` then getting passed here where we create a `dataclass` from that then creating another class inheriting from that and `Streamable` (again) and then we always decorate with `dataclass` (again) after this as I recall...

My thinking was that we want this applied to a `Streamable` subclass so hinting that would encourage people to not apply it without having inherited from `Streamable`.  Perhaps with proper type checks we would even become comfortable removing the inheritance that is applied below?  I dunno, I should probably just not delve into all this now.  Maybe just an `isinstance()` with an error instead of reapplying the inheritance in this function?

So, back to what to hint, if anything.  As noted at the end of the docstring below, I think we want people to have applied the `Streamable` inheritance themselves so I think we would rather hint as it is to encourage that?

Reply:
That does sound right that when we've created a new class intended to be streamable, we both subclass and decorate it. But maybe we shouldn't have been doing that?